src/contract/subcontracts/base.ts

```diff
@@ -1,246 +1,97 @@
-   * @category Internal Contract
-   */
-  get peripheryContract(): PeripheryAbi {
-    return this.getPeripheryContract(this._signerOrProvider)
-  }
-
-  protected getPeripheryContract(
-    signerOrProvider:
-      | ethers.Signer
-      | ethers.providers.Provider = this.getDefaultProvider(),
-  ) {
-    return PeripheryAbi__factory.connect(
-      this.options.peripheryContractAddress,
-      signerOrProvider,
-    )
-  }
-
-  /**
-   * Returns the internal cbt contract.
-   * @category Internal Contract
-   */
-  get cbtContract(): CbtAbi {
-    return this.getCbtContract(this._signerOrProvider)
-  }
-
-  protected getCbtContract(
-    signerOrProvider:
-      | ethers.Signer
-      | ethers.providers.Provider = this.getDefaultProvider(),
-  ) {
-    return CbtAbi__factory.connect(
-      this.options.cbtContractAddress,
-      signerOrProvider,
-    )
-  }
-
-  /**
-   * Returns the internal cbt contract.
-   */
-  get tipsContract(): TipsAbi {
-    return this.getTipsContract(this._signerOrProvider)
-  }
-
-  protected getTipsContract(
-    signerOrProvider:
-      | ethers.Signer
-      | ethers.providers.Provider = this.getDefaultProvider(),
-  ) {
-    return TipsAbi__factory.connect(
-      this.options.tipsContractAddress,
-      signerOrProvider,
-    )
-  }
-
-  /**
-   * Returns the internal mira contract.
-   */
-  get miraContract(): MiraAbi {
-    return this.getMiraContract(this._signerOrProvider)
-  }
-
-  protected getMiraContract(
-    signerOrProvider:
-      | ethers.Signer
-      | ethers.providers.Provider = this.getDefaultProvider(),
-  ) {
-    return MiraAbi__factory.connect(
-      this.options.miraContractAddress,
-      signerOrProvider,
-    )
-  }
-
-  /**
-   * Returns the internal linklist contract.
-   */
-  get linklistContract(): LinklistAbi {
-    return this.getLinklistContract(this._signerOrProvider)
-  }
-
-  protected getLinklistContract(
-    signerOrProvider:
-      | ethers.Signer
-      | ethers.providers.Provider = this.getDefaultProvider(),
-  ) {
-    return LinklistAbi__factory.connect(
-      this.options.linklistContractAddress,
-      signerOrProvider,
-    )
-  }
-
-  /**
-   * Returns the internal newbie villa contract.
-   * @category Internal Contract
-   */
-  get newbieVillaContract(): NewbieVillaAbi {
-    return this.getNewbieVillaContract(this._signerOrProvider)
-  }
+export type ResolvedContractOptions = Overwrite<
+  ContractOptions,
+  { address: AddressOptions }
+>
 
-  protected getNewbieVillaContract(
-    signerOrProvider:
-      | ethers.Signer
-      | ethers.providers.Provider = this.getDefaultProvider(),
-  ) {
-    return NewbieVillaAbi__factory.connect(
-      this.options.newbieVillaContractAddress,
-      signerOrProvider,
-    )
-  }
+export class BaseContract {
+  publicClient: PublicClient = createDefaultPublicClient()
+  walletClient: WalletClient<Transport, Chain, Account> | undefined
+  #account: Address | Account | undefined
```

Comment:
I think we can narrow this type by making the `BaseContract` a generic class?

Reply:
It's a bit hard. I would like to try in annother PR.